File: pt-1/3876.construct-uniform-parity-array-ii.go
package leetcode

//
// @lc app=leetcode id=3876 lang=golang
//
// [3876] Construct Uniform Parity Array II
//
// Notes
// If all numbers have the same parity, the array is already uniform. Otherwise
// the Python criterion is exactly whether the minimum odd value is smaller than
// the minimum even value. We track both minima in one pass. Time: O(n). Space:
// O(1).
//
// @lc code=start

func UniformArray3876(nums1 []int) bool {
	const inf = 1 << 60
	minOdd := inf
	minEven := inf

	for _, value := range nums1 {
		if value%2 != 0 {
			if value < minOdd {
				minOdd = value
			}
		} else if value < minEven {
			minEven = value
		}
	}

	if minOdd == inf || minEven == inf {
		return true
	}
	return minOdd < minEven
}

// @lc code=end

File: pt-1/3876.construct-uniform-parity-array-ii.java
/*
 * @lc app=leetcode id=3876 lang=java
 *
 * [3876] Construct Uniform Parity Array II
 *
 * If all values already have the same parity, success is immediate. Otherwise,
 * the Python logic checks whether the smallest odd value is smaller than the
 * smallest even value.
 *
 * Time: O(n). Space: O(1).
 */

// @lc code=start
class Solution {
    public boolean uniformArray(int[] nums1) {
        int minOdd = Integer.MAX_VALUE;
        int minEven = Integer.MAX_VALUE;

        for (int value : nums1) {
            if ((value & 1) == 1) {
                minOdd = Math.min(minOdd, value);
            } else {
                minEven = Math.min(minEven, value);
            }
        }
        if (minOdd == Integer.MAX_VALUE || minEven == Integer.MAX_VALUE) {
            return true;
        }
        return minOdd < minEven;
    }
}
// @lc code=end
